### ISSUE TITLE
Delegation pattern Exercise #4

### ISSUE THREAD

**Issue by zandratay** [nus-cs2030/2223-s1#55]
Hi, I was wondering how we should approach the add function of exercise 4.
i tried both approaches in line 20 and line 21 as shown below:

 19  ```
Grid2D<E> add(E elem) {
 20         Grid2D<E> newGrid = new Grid2D(new ImList<E>(this.list), this.numOfCols);
 21         ImList<E> newList = new ImList<E>(this.list);
 22         // other processes
 23     }

```
but kept running into the error: `ImList<E> cannot be converted to List<E>`. Yet, we are required to use List in our constructor: `Grid2D(List<E> list, int numOfCols)`. Is this an issue with my syntax or something else?

**Comment by shirohaku** [comment]
Not sure if this method is the answer they want us to arrive at, but I used another private overloaded constructor to make Grid2D take in an ImList instead of a List.

**Comment by bryanolm99** [comment]
I don't think we can create a new ImList by using this.list as a parameter, because the ImList constructor only accepts lists that implements java's List interface (e.g. ArrayList, LinkedList). In our case, this.list is of type ImList, which does not implement Java's List interface.

**Comment by zandratay** [comment]
> Not sure if this method is the answer they want us to arrive at, but I used another private overloaded constructor to resolve this

As in, u constructed Grid2D with ImList<E> and numOfCols as the parameter? 
What about the original constructor as specified by the question, that was referring to a list?

**Comment by zandratay** [comment]
> > 20         Grid2D newGrid = new Grid2D(new ImList(this.list), this.numOfCols);
> > 21         ImList newList = new ImList(this.list);
> 
> I don't think we can create a new ImList by using this.list as a parameter, because the ImList constructor only accepts lists that implements java's List interface (e.g. ArrayList, LinkedList). In our case, this.list is of type ImList, which does not implement Java's List interface.

Oh thank you that helped clear things up a lot! In this case, how are we even supposed to create a new ImList object to make use of its methods though? Sorry… I am very lost at this exercise

**Comment by chiralcentre** [comment]
You can use an overloaded constructor which takes in ImList as the parameters to help circumvent your issue
